### PR TITLE
Syntaxerror in variable unpacking 487

### LINKED ISSUE #487: SyntaxError in variable unpacking
```Python
for x, *y in (enumerate(range(5))): pass
```
> SyntaxError: invalid syntax

### PR DESCRIPTION
**Pull request by hackf5**
Resolves #487

as mentioned in the chat I've tried to minimize the changes, however I chose to make a deoptimization in FinishAssignments so that it always allocates a list at the start. imo the improved readability and reduction in cyclomatic complexity outweigh the cost of the allocation. see what you think.

probably won't be able to make any updates for the next week and a half due to other commitments.

### CODE REVIEW

File: Src/IronPython/Compiler/Parser.cs [hunk 1]
```diff
@@ -2710,31 +2742,51 @@ private Parameter ParseDefParameter(HashSet<string> names, ParameterKind paramet
...
             var oEnd = GetEnd();
-            int grouping = _tokenizer.GroupingLevel;
+            var grouping = _tokenizer.GroupingLevel;
 
             Expression ret;
             if (MaybeEat(TokenKind.RightBracket)) {
                 ret = new ListExpression();
-            } else {
-                bool prevAllow = _allowIncomplete;
+            }
```
Comment (slozier) [inline Src/IronPython/Compiler/Parser.cs]: Just a note on style: `} else {`

If you're using VS then formatting the document should pick up the editorconfig and apply the correct style throughout.

File: Src/IronPython/Runtime/Operations/PythonOps.cs [hunk 2]
```diff
@@ -3812,7 +3812,7 @@ public static partial class PythonOps {
             }
             else {
                 Debug.Assert(right < left + argcntafter);
-                return ValueError("not enough values to unpack (expected at least {0}, got {1})", left + argcntafter, right);
```
[the remark this answers is not in the document]
Comment (hackf5) [reply Src/IronPython/Runtime/Operations/PythonOps.cs]: done

File: Src/IronPython/Compiler/Ast/SequenceExpression.cs [hunk 3]
```diff
@@ -136,10 +136,24 @@ public abstract class SequenceExpression : Expression {
...
-            foreach (var item in _items) {
-                var res = item.CheckAssign();
-                if (res != null) return res;
+            var starCount = 0;
+            foreach (var item in Items) {
+                if (item.CheckAssign() != null) {
+                    // we don't return the same message here as CPython doesn't seem to either, 
+                    // for example ((yield a), 2,3) = (2,3,4) gives a different error than
+                    // a = yield 3 = yield 4.
+                    return "can't assign to " + item.NodeName;
```
[the remark this answers is not in the document]
Comment (hackf5) [reply Src/IronPython/Compiler/Ast/SequenceExpression.cs]: done.

The error message was taken from in TupleExpression, which is why it was written like this.

File: Src/IronPython/Compiler/Parser.cs [hunk 4]
```diff
@@ -2130,23 +2152,23 @@ private Parameter ParseDefParameter(HashSet<string> names, ParameterKind paramet
         }
 
         // exprlist: (expr|star_expr) (',' (expr|star_expr))* [',']
-        // TODO: handle star_expr
         private List<Expression> ParseExprList(out bool trailingComma) {
-            List<Expression> l = new List<Expression>();
-            trailingComma = false;
+            var expressions = new List<Expression>();
```
[the remark this answers is not in the document]
Comment (hackf5) [reply Src/IronPython/Compiler/Parser.cs]: done, test added.

File: Src/IronPython/Compiler/Parser.cs [hunk 5]
```diff
@@ -2318,50 +2340,60 @@ private Parameter ParseDefParameter(HashSet<string> names, ParameterKind paramet
             }
         }
 
+        // star_expr
```
[the remark this answers is not in the document]
Comment (hackf5) [reply Src/IronPython/Compiler/Parser.cs]: done.

File: Src/IronPython/Compiler/Parser.cs [hunk 6]
```diff
@@ -2318,50 +2340,60 @@ private Parameter ParseDefParameter(HashSet<string> names, ParameterKind paramet
...
-                l.Add(expr);
+            if (!MaybeEat(TokenKind.Comma)) {
+                // there is no comma so that is the end of the list.
+                return expr;
+            }
+
+            // the presence of a comma indicates that the list may contain more elements
+            return ParseTestListStarExprTail();
+
+            Expression ParseTestListStarExprTail() {
```
[the remark this answers is not in the document]
Comment (hackf5) [reply Src/IronPython/Compiler/Parser.cs]: done.

File: Src/IronPython/Compiler/Parser.cs (hunk 1 above)
Comment (hackf5) [reply Src/IronPython/Compiler/Parser.cs]: I've got VS and also ReSharper. My auto-format is currently doing too much as it is setup to match the StyleCop format, however closing an outer brace works.

I could take a look at formalizing the coding standard with some analyzers if you are interested?

File: Src/IronPython/Compiler/Parser.cs [hunk 7]
```diff
@@ -613,45 +624,31 @@ public class Parser : IDisposable { // TODO: remove IDisposable
...
 
         private Statement FinishAssignments(Expression right) {
-            List<Expression> left = null;
-            Expression singleLeft = null;
+            var left = new List<Expression>();
 
             while (MaybeEat(TokenKind.Assign)) {
-                string assignError = right.CheckAssign();
-                if (assignError != null) {
+                if (right.CheckAssign() is { } assignError) {
```
[the remark this answers is not in the document]
Comment (hackf5) [reply Src/IronPython/Compiler/Parser.cs]: It was a new one for me too, it was a ReSharper suggestion.

File: Src/IronPython/Compiler/Parser.cs [hunk 8]
```diff
@@ -2710,31 +2742,51 @@ private Parameter ParseDefParameter(HashSet<string> names, ParameterKind paramet
...
+                    if (PeekToken(Tokens.KeywordForToken)) {
+                        // although it's calling ParseCompIter(), because the peek token is a FOR it is going to
+                        // do the right thing.
+                        ret = new ListComprehension(expr, ParseCompIter());
                     } else {
-                        Eat(TokenKind.RightBracket);
-                        ret = new ListExpression(t0);
+                        // (',' (test|star_expr))* [',']
+                        var items = new List<Expression> { expr };
+                        while (MaybeEat(TokenKind.Comma)) {
```
[the remark this answers is not in the document]
Comment (hackf5) [reply Src/IronPython/Compiler/Parser.cs]: hmmm. it looks as though there is sufficient difference in the way that the comma is handled that the duplication is probably necessary. I've inlined `ParseTestListStarExprTail` though as I think this makes the differences more explicit.

File: Src/IronPython/Compiler/Parser.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (hackf5) [reply Src/IronPython/Compiler/Parser.cs]: actually, based on one of your other comments I've inlined this method.

File: Src/IronPython/Compiler/Parser.cs [hunk 9]
```diff
@@ -613,45 +624,31 @@ public class Parser : IDisposable { // TODO: remove IDisposable
         }
 
         private Statement FinishAssignments(Expression right) {
-            List<Expression> left = null;
```
[the remark this answers is not in the document]
Comment (hackf5) [reply Src/IronPython/Compiler/Parser.cs]: My main issue was with the duplication at the end of the method. I think I've managed to restore the efficiency, but remove the duplication.

File: Src/IronPython/Compiler/Parser.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (hackf5) [reply Src/IronPython/Compiler/Parser.cs]: While I was looking at this I found that the tuple path also needed updating however, so I've fixed this and added some tests.

File: Src/IronPython/Compiler/Parser.cs [hunk 10]
```diff
@@ -635,13 +636,22 @@ public class Parser : IDisposable { // TODO: remove IDisposable
...
+                    left.Add(right);
+                }
+
                 right = MaybeEat(TokenKind.KeywordYield) ? ParseYieldExpression() : ParseTestListStarExpr();
             }
 
             CheckNotAssignmentTargetOnly(right);
 
-            var target = left.ToArray();
+            var target = left?.ToArray() ?? new [] { singleLeft };
```
Comment (slozier) [inline Src/IronPython/Compiler/Parser.cs]: Looking at the method without context it wasn't clear that `singleLeft` is never null at this point. I wonder if it's worth adding `Debug.Assert(singleLeft != null)`. Maybe also `Debug.Assert(PeekToken(TokenKind.Assign))` at the top of the method?

File: Src/IronPython/Compiler/Parser.cs (hunk 1 above)
Comment (slozier) [reply Src/IronPython/Compiler/Parser.cs]: Is there an analyzer that picks up the rules from editorconfig? Some parts of the codebase (e.g. BZip2 and c#sqlite) have their own inconsistent style (although I haven't set up an editorconfig for those).

**Comment by slozier** [comment]
@hackf5 Any other plans for this PR? Otherwise I'm happy with it and will merge it in.

File: Src/IronPython/Compiler/Parser.cs (hunk 1 above)
Comment (master-minszor) [reply Src/IronPython/Compiler/Parser.cs]: No, but VS with ReSharper should pick it up.

**Comment by slozier** [comment]
Thanks for the PR!

**Comment by master-minszor** [comment]
You're welcome!

**Pull request merged**